libcaja-private/caja-dnd.c

```diff
@@ -517,6 +517,12 @@ caja_drag_default_drop_action_for_icons (GdkDragContext *context,
     else
     {
         target = g_file_new_for_uri (target_uri_string);
+
+        if (target_file == NULL)
+        {
+            target_file = caja_file_get_by_uri (target_uri_string);
```

Comment:
You called twice `caja_file_get_internal`:
1. `caja_file_get_internal (location, FALSE); ` on `target_file = caja_file_get_existing_by_uri (target_uri_string);`
2. `caja_file_get_internal (location, TRUE);` on `target_file = caja_file_get_by_uri (target_uri_string);`.

So, if the directory doesn't exist in the directories lookup table (1), you add it on (2).

```C
CajaFile *
caja_file_get_existing_by_uri (const char *uri)
{
        GFile *location;
        CajaFile *file;

        location = g_file_new_for_uri (uri);
        file = caja_file_get_internal (location, FALSE);
        g_object_unref (location);

        return file;
}

CajaFile *
caja_file_get_by_uri (const char *uri)
{
        GFile *location;
        CajaFile *file;

        location = g_file_new_for_uri (uri);
        file = caja_file_get_internal (location, TRUE);
        g_object_unref (location);

        return file;
}
```

```C
CajaDirectory *
caja_directory_get_internal (GFile *location, gboolean create)
{
    CajaDirectory *directory;

    /* Create the hash table first time through. */
    if (directories == NULL) {
        directories = g_hash_table_new (g_file_hash, (GCompareFunc) g_file_equal);
        caja_global_preferences_init ();
    }

    /* If the object is already in the hash table, look it up. */

    directory = g_hash_table_lookup (directories,
                                     location);
    if (directory != NULL)
    {
        caja_directory_ref (directory);
    }
    else if (create)
    {
        /* Create a new directory object instead. */
        directory = caja_directory_new (location);
        if (directory == NULL)
        {
            return NULL;
        }

        /* Put it in the hash table. */
        g_hash_table_insert (directories,
                             directory->details->location,
                             directory);
    }

    return directory;
}
```

Reply:
What is the purpose of `caja_file_refresh_info (target_file);`?

Reply:
we may be using two entries for the same location:
![VirtualBox_11 Fedora Rawhide_27_02_2021_21_06_22](https://user-images.githubusercontent.com/10171411/109399008-6d16a300-7940-11eb-992b-160c1b2cda16.png)

Reply:
My first fix attempt was replacing caja_file_get_existing_by_uri with caja_file_get_by_uri on caja-dnd.c:458.

```diff
- target_file = caja_file_get_existing_by_uri (target_uri_string);
+ target_file = caja_file_get_by_uri (target_uri_string);
```
This solves target_file being null, but then target_file->details->filesystem_id is null. So you still get the same result in check_same_fs.

The purpose of caja_file_refresh_info is to load the right value in filesystem_id. I had a look at the existing functions and that was the most appropriate for the task that I could find, but note that I'm not familiar with the code of caja.

I'm aware that caja_file_get_internal is being called twice. I left it as is in order to keep the fix as simple and less intrusive as possible. You could also revert to the original replacement on line 458 and then use a different criteria to determine if it's necessary to update filesystem_id or not.

Reply:
Maybe the issue is due the actual target does not exist in the lookup table of directories because the virtual file system uri was added instead, so it failed to check whether the source and target belong to the same file system, which is why the copy operation is selected instead of the move operation.

Reply:
You know the internals of caja far better than me, please feel free to solve the issue in any other way that you consider appropriate. Just tell me if there's any way I can help in order to have the problem solved asap, because I find it annoying. Thank you very much.